metagraph/src/annotation/taxonomy/taxonomic_db.cpp

```diff
@@ -0,0 +1,326 @@
+    logger->trace("Finished parsing taxonomic tree in '{}' sec", timer.elapsed());
+
+    timer.reset();
+    logger->trace("Calculating tree statistics..");
+    std::vector<NormalizedTaxId> tree_linearization;
+    node_to_linearization_idx.resize(tree.size());
+    node_depth.resize(tree.size());
+    dfs_statistics(root_node, tree, tree_linearization);
+    logger->trace("Finished calculating tree statistics in '{}' sec", timer.elapsed());
+
+    timer.reset();
+    logger->trace("Starting rmq preprocessing..");
+    rmq_preprocessing(tree_linearization);
+    logger->trace("Finished rmq preprocessing in '{}' sec", timer.elapsed());
+}
+
+NormalizedTaxId TaxonomyDB::find_lca(const std::vector<NormalizedTaxId> &taxids) const {
+    if (taxids.empty()) {
+        logger->error("Can't find LCA for an empty set of taxids.");
+        std::exit(1);
+    }
+    uint64_t left_idx = node_to_linearization_idx[taxids[0]];
+    uint64_t right_idx = node_to_linearization_idx[taxids[0]];
+    for (const auto &taxid: taxids) {
+        if (node_to_linearization_idx[taxid] < left_idx) {
+            left_idx = node_to_linearization_idx[taxid];
+        }
+        if (node_to_linearization_idx[taxid] > right_idx) {
+            right_idx = node_to_linearization_idx[taxid];
+        }
+    }
+
+    uint64_t log_dist = precalc_log2[right_idx - left_idx];
+    uint64_t left_lca = rmq_data[log_dist][left_idx];
+    uint64_t right_lca = rmq_data[log_dist][right_idx - precalc_pow2[log_dist] + 1];
+
+    if (node_depth[left_lca] > node_depth[right_lca]) {
+        return left_lca;
+    }
+    return right_lca;
+}
+
+bool TaxonomyDB::get_normalized_taxid(const std::string accession_version, NormalizedTaxId &taxid) const {
+    num_get_taxid_calls += 1;
+    if (! lookup_table.count(accession_version)) {
+        // accession_version not in the lookup_table
+        num_get_taxid_calls_failed += 1;
+        return false;
+    }
+    if (! normalized_taxid.count(lookup_table.at(accession_version))) {
+        // taxid (corresponding to the current accession_version) not in the taxonomic tree.
+        return false;
+    }
+
+    taxid = normalized_taxid.at(lookup_table.at(accession_version));
+    return true;
+}
+
+void TaxonomyDB::kmer_to_taxid_map_update(const annot::MultiLabelEncoded<std::string> &annot,
+                                          std::mutex &taxo_mutex) {
+    for (const auto &label: annot.get_all_labels()) {
+        AccessionVersion accession_version = get_accession_version_from_label(label);
+        uint64_t taxid;
+        if(!get_normalized_taxid(accession_version, taxid)) {
+            continue;
+        }
+        annot.call_objects(label, [&](const auto &index) {
+          if (index <= 0) {
+              return;
+          }
+          if (index >= taxonomic_map.size()) {
+              // Double the size of taxonomic_map until the current 'index' fits inside.
+              taxo_mutex.lock();
```

Comment:
Why do you need locks? What is running concurrently? (It's hard to see it from this code)

Reply:
The locks are only for the computation of the LCA, which is O(1). Lets discuss at the meeting if it is worth keeping it or not. Thank you very much for the review!